File: backend/app/schemas/ai_selection.py
"""AI 选品 API 请求与响应模型。"""

from typing import List, Optional

from pydantic import BaseModel, Field

MAX_ASINS_PER_REQUEST = 200


class AsinLookupRequest(BaseModel):
    """ASIN 查询请求。"""

    asins: List[str] = Field(
        ..., min_length=1, max_length=MAX_ASINS_PER_REQUEST
    )
    marketplace: Optional[str] = Field(None, pattern="^(UK|DE|US)$")


class PushRequest(AsinLookupRequest):
    """AI Agent 投递请求。"""

    message: Optional[str] = Field(None, max_length=200)


class AiSelectionProduct(BaseModel):
    """兼容 UniversalCard 的商品快照。"""

    asin: str = ""
    productTitle: str = ""
    imageUrl: str = ""
    price: Optional[float] = None
    marketplace: str = ""
    bsr: Optional[int] = None
    units: Optional[int] = None
    unitsGr: Optional[float] = None
    sellerName: str = ""
    mainCategoryName: str = ""
    availableDate: Optional[int] = None
    createdAt: Optional[str] = None
    symbol: str = ""
    rating: Optional[float] = None
    ratings: Optional[int] = None
    fulfillment: str = ""
    sourceTable: str = ""
    productUrl: str = ""
    listingDays: Optional[int] = None


class PushBatch(BaseModel):
    """一次投递批次。"""

    id: str
    message: str = ""
    pushedAt: str
    total: int
    requested: int = 0
    invalidAsins: List[str] = Field(default_factory=list)
    products: List[AiSelectionProduct] = Field(default_factory=list)


class PushResponse(BaseModel):
    batchId: str
    total: int
    requested: int
    invalidAsins: List[str] = Field(default_factory=list)
    products: List[AiSelectionProduct] = Field(default_factory=list)
    message: str


class SessionResponse(BaseModel):
    batches: List[PushBatch] = Field(default_factory=list)


class AsinLookupResponse(BaseModel):
    total: int
    requested: int = 0
    invalidAsins: List[str] = Field(default_factory=list)
    products: List[AiSelectionProduct] = Field(default_factory=list)


class AutoScreeningConfig(BaseModel):
    enabled: bool = False
    method_cards: List[str] = Field(default_factory=lambda: ["M01", "M03"])


class AutoScreeningResponse(BaseModel):
    total_raw: int = 0
    total_after_screening: int = 0
    products: List[AiSelectionProduct] = Field(default_factory=list)
    message: str = "自动筛选框架已预留，暂未实现"

File: frontend/src/api/ai-selection.ts
import request from "@/utils/request";

const BASE = "/api/v1/ai-selection";

/** AI 选品产品（后端 camelCase，直接兼容 UniversalCard） */
export interface AiSelectionProduct {
  asin: string;
  productTitle: string;
  imageUrl: string;
  price: number | null;
  marketplace: string;
  bsr: number | null;
  units: number | null;
  unitsGr: number | null;
  sellerName: string;
  mainCategoryName: string;
  availableDate: number | null;
  createdAt: string | null;
  symbol: string;
  rating: number | null;
  ratings: number | null;
  fulfillment: string;
  sourceTable: string;
  productUrl: string;
  listingDays: number | null;
}

export interface PushBatch {
  id: string;
  message: string;
  pushedAt: string;
  total: number;
  requested: number;
  invalidAsins: string[];
  products: AiSelectionProduct[];
}

export interface SessionResponse {
  batches: PushBatch[];
}

export interface PushResponse {
  batchId: string;
  total: number;
  requested: number;
  invalidAsins: string[];
  products: AiSelectionProduct[];
  message: string;
}

export interface AsinLookupResponse {
  total: number;
  requested: number;
  invalidAsins: string[];
  products: AiSelectionProduct[];
}

// 响应拦截器已解包为业务对象，这里直接把结果断言为对应结构。
function unwrap<T>(p: Promise<unknown>): Promise<T> {
  return p.then((res) => res as T);
}

/** AI Agent 投递 ASIN — 调 API 后前端自动轮询展示 */
export function pushAsins(
  asins: string[],
  options?: { marketplace?: string; message?: string },
): Promise<PushResponse> {
  return unwrap<PushResponse>(
    request.post(`${BASE}/push`, {
      asins,
      marketplace: options?.marketplace,
      message: options?.message,
    }),
  );
}

/** 前端轮询 — 首次取最近批次，后续只取指定批次之后的增量 */
export function getSession(
  afterBatchId?: string,
  limit = 10,
): Promise<SessionResponse> {
  return unwrap<SessionResponse>(
    request.get(`${BASE}/session`, {
      params: { afterBatchId, limit },
    }),
  );
}

/** 清空当前用户会话 */
export function clearSession(): Promise<{ message: string }> {
  return unwrap<{ message: string }>(request.delete(`${BASE}/session`));
}

/** 一次性 ASIN 查询（不入会话） */
export function lookupAsins(
  asins: string[],
  marketplace?: string,
): Promise<AsinLookupResponse> {
  return unwrap<AsinLookupResponse>(
    request.post(`${BASE}/lookup`, { asins, marketplace }),
  );
}
